server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,279 @@
+import org.opensearch.index.store.remote.utils.filetracker.FileState;
+import org.opensearch.index.store.remote.utils.filetracker.FileType;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class CompositeDirectory extends FilterDirectory {
+    private static final Logger logger = LogManager.getLogger(CompositeDirectory.class);
+
+    private final FSDirectory localDirectory;
+    private final RemoteStoreFileTrackerAdapter remoteStoreFileTrackerAdapter;
+    private final FileCache fileCache;
+    private final AtomicBoolean isRemoteDirectorySet;
+    private final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final ReentrantReadWriteLock.WriteLock writeLock = readWriteLock.writeLock();
+    private final ReentrantReadWriteLock.ReadLock readLock = readWriteLock.readLock();
+
+    public CompositeDirectory(FSDirectory localDirectory, FileCache fileCache) {
+        super(localDirectory);
+        this.localDirectory = localDirectory;
+        this.fileCache = fileCache;
+        this.remoteStoreFileTrackerAdapter = new CompositeDirectoryRemoteStoreFileTrackerAdapter(fileCache);
+        isRemoteDirectorySet = new AtomicBoolean(false);
+    }
+
+    public void setRemoteDirectory(Directory remoteDirectory) {
+        logger.trace("Setting remote Directory ...");
+        if (!isRemoteDirectorySet()) {
+            ((CompositeDirectoryRemoteStoreFileTrackerAdapter) remoteStoreFileTrackerAdapter).setRemoteDirectory(remoteDirectory);
+            isRemoteDirectorySet.set(true);
+        }
+    }
+
+    @Override
+    public String[] listAll() throws IOException {
+        logger.trace("listAll() called ...");
+        readLock.lock();
+        try {
+            String[] remoteFiles = new String[0];
+            String[] localFiles = localDirectory.listAll();
+            if (isRemoteDirectorySet()) remoteFiles = ((CompositeDirectoryRemoteStoreFileTrackerAdapter) remoteStoreFileTrackerAdapter)
+                .getRemoteFiles();
+            logger.trace("LocalDirectory files : " + Arrays.toString(localFiles));
+            logger.trace("Remote Directory files : " + Arrays.toString(remoteFiles));
+            Set<String> allFiles = new HashSet<>(Arrays.asList(localFiles));
+            allFiles.addAll(Arrays.asList(remoteFiles));
+
+            Set<String> localLuceneFiles = allFiles.stream().filter(file -> !isBlockFile(file)).collect(Collectors.toUnmodifiableSet());
+            String[] files = new String[localLuceneFiles.size()];
+            localLuceneFiles.toArray(files);
+            Arrays.sort(files);
+
+            logger.trace("listAll() returns : " + Arrays.toString(files));
```

Comment:
Use a lazy implementation of the logger, this will cause unnecessary computation

Reply:
Done.